OpenTrons/opentrons-api
# refactor(robot-server, system-server): commonize server code

# Overview

With the addition of the system server, we now have two servers with very similar architectures in our robot stack; both the robot-server and the system-server provide a FastAPI-backed API and use SQLite via `sqlalchemy`. It's expected that both servers will have to perform very similar things, which means they can benefit from having a place to share common code. 

This PR adds a `server-utils` package. This is a Python package that is imported by both the `robot-server` and `system-server`. For now, this PR only moves over code that is _completely trivial_ to port - most of the changes should just be changing import paths.

On the OT-2, this just goes into the same directory as all of the other python packages. For the OT-3, openembedded is able to pick up the `pipfile` requirements from `robot-server` and `system-server` and a copy of this is saved locally to each server's folder in `/opt/`. The makefile pushes to both of these destinations accordingly.

# Test Plan

There is a companion PR in the `buildroot` repo with the necessary changes to add this to the OT-2, but the OT-3 build should work from `oe-core`'s edge.

To test:
* Install a fresh build that includes `server-utils`
* Verify that the robot server is running by connecting to the robot via the app. Send a protocol, generally exercise it a bit to make sure it's working correctly.
* `ssh` in and verify with `systemctl` that the system server is running as well
* Try removing the `server-utils` package and `make push` it back; verify that it is uploaded to the correct place on the robot, and restarting each server works correctly.

Tested on:
* [x] OT-2
* [x] OT-3

# Changelog

* Added a `server-utils` Python project, and added as a dependency to both `robot-server` and `system-server`
* Moved `robot_server.util` to `server-utils`
* Moved `robot_server.app_state` to `server-utils`
* Updated `g-code-tests` to import `server-utils`
* Added `server-utils` to appropriate buildroot files

# Review requests

There are a lot of files being touched in this PR, but _most_ of it is just changing `import` statements around. 

* Is any of this _not_ appropriate to commonize?
* How in-depth should the robot server testing be? Is uploading a simple protocol + running it sufficient?

# Risk assessment

Medium! The actual code changes are minimal in `robot-server`, but if there's any misconfiguration in imports then the code won't work at all. It should be easy to catch any issues with smoke testing at least.
